src/osd/scrubber/scrub_queue.cc

```diff
@@ -0,0 +1,639 @@
+	   << dendl;
+  ceph_assert(t.urgency > urgency_t::off);
+  std::unique_lock l{jobs_lock};
+  t.is_valid = true;
+  to_scrub.push_back(t);
+}
+
+void ScrubQueue::dump_scrubs(ceph::Formatter* f)
+{
+  std::lock_guard lck(jobs_lock);
+  normalize_the_queue();
+
+  f->open_array_section("scrubs");
+  std::for_each(to_scrub.cbegin(), to_scrub.cend(), [&f](const auto& j) {
+    j.dump("sched-target", f);
+  });
+  f->close_section();
+}
+
+
+// ////////////////////////////////////////////////////////////////////////// //
+// initiating a scrub
+
+using ScrubPreconds = Scrub::ScrubPreconds;
+using schedule_result_t = Scrub::schedule_result_t;
+
+void ScrubQueue::sched_scrub(
+    const ceph::common::ConfigProxy& config,
+    bool is_recovery_active)
+{
+  utime_t scrub_tick_time = scrub_clock_now();
+  dout(10) << fmt::format(
+		  "time now:{}, is_recovery_active:{}", scrub_tick_time,
+		  is_recovery_active)
+	   << dendl;
+
+  // do the OSD-wide environment conditions, and the availability of scrub
+  // resources, allow us to start a scrub?
+  auto maybe_env_cond =
+      preconditions_to_scrubbing(config, is_recovery_active, scrub_tick_time);
+  if (!maybe_env_cond) {
+    return;
+  }
+  auto preconds = maybe_env_cond.value();
+
+  std::unique_lock l{jobs_lock};
+
+  // partition and sort the queue
+  if (bool not_empty = normalize_the_queue(); !not_empty) {
+    dout(10) << fmt::format("{}: no eligible scrub targets", __func__) << dendl;
+    return;
+  }
+
+  // pop the first job from the queue, as a candidate
+  auto cand = to_scrub.front();
+  to_scrub.pop_front();
+  l.unlock();
+
+  PgLockWrapper locked_g = osd_service.get_locked_pg(cand.pgid);
+  PGRef pg = locked_g.m_pg;
+  if (!pg) {
+    // the PG was deleted in the sort time since unlocking the queue
+    dout(5) << fmt::format("{}: pg[{}] not found", __func__, cand.pgid)
+	    << dendl;
+    return;
+  }
+  pg->start_scrubbing(scrub_tick_time, cand.level, preconds);
+}
+
+
+tl::expected<ScrubPreconds, schedule_result_t>
+ScrubQueue::preconditions_to_scrubbing(
+    const ceph::common::ConfigProxy& config,
+    bool is_recovery_active,
+    utime_t scrub_clock_now)
+{
+  if (auto blocked_pgs = get_blocked_pgs_count(); blocked_pgs > 0) {
+    // some PGs managed by this OSD were blocked by a locked object during
+    // scrub. This means we might not have the resources needed to scrub now.
+    dout(10) << fmt::format(
+		    "{}: PGs are blocked while scrubbing due to locked objects "
+		    "({} PGs)",
+		    __func__, blocked_pgs)
+	     << dendl;
+  }
+
+  // sometimes we just skip the scrubbing
+  if ((rand() / (double)RAND_MAX) < config->osd_scrub_backoff_ratio) {
+    dout(20) << fmt::format(
+		    "{}: lost coin flip, randomly backing off (ratio: {:f})",
+		    __func__, config->osd_scrub_backoff_ratio)
+	     << dendl;
+    return tl::unexpected(schedule_result_t::failure);
+  }
+
+  // fail fast if no resources are available
+  if (!can_inc_scrubs()) {
```

Comment:
This also seems really orthogonal to the problem of ordering a set of items.  Why not do this in the OSD entry point?